src/geom/SpanSet.cc

```diff
@@ -99,6 +99,72 @@ bool spansContiguous(Span const& a, Span const& b, bool compareY = true) {
                    : false;
 }
 
+/* Determine the intersection with a mask or its logical inverse
+ *
+ * spanSet - SpanSet object with which to intersect the mask
+ * mask - Mask object to be intersectedNot
+ * bitmask - bitpattern to used when deciding pixel membership in the mask
+ *
+ * Templates:
+ * T - pixel type of the mask
+ * invert - boolean indicating if the intersection is with the mask or the inverse of the mask
+            templated parameter to ensure the check is compiled away if it can be
+*/
+template <typename T, bool invert>
+std::shared_ptr<SpanSet> maskIntersect(SpanSet const & spanSet, image::Mask<T> const & mask, T bitmask) {
+    // This vector will store our output spans
+    std::vector<Span> newVec;
+    // Variable to mark if a new span has been started
+    bool started = false;
+    auto maskBBox = mask.getBBox();
+    for (auto const & spn : spanSet) {
+        // Reset the started flag for each different span
+        started = false;
+        int y = spn.getY();
+        // Reset the min and max variables
+        int minX = 0;
+        int maxX = 0;
+        for (int x = spn.getMinX(); x <= spn.getMaxX(); ++x){
+            // Verify that the point actually falls within the bounds of the mask
+            if (maskBBox.contains(Point2I(x, y))) {
+                // Find if the pixel matches the given bit pattern
+                bool pixelCompare = mask.get0(x, y) & bitmask;
+                // if the templated boolean indicates the compliment of the mask is desired, invert the
+                // pixel comparison
+                if (invert) {
+                    pixelCompare = !pixelCompare;
+                }
+                // If the pixel is to be included in the operation, start or append recording values
+                // needed to later construct a Span
+                if (pixelCompare) {
+                    if (!started) {
+                        started = true;
+                        minX = x;
+                        maxX = x;
+                    } else {
+                        if (minX > x) {
+                            minX = x;
+                        } else if (maxX < x) {
+                            maxX = x;
+                        }
+                    }
+                    if (x == spn.getMaxX()) {
+                        // We have reached the end of a row and the pixel evaluates true, should add the
+                        // Span
+                        newVec.push_back(Span(y, minX, maxX));
+                    }
+                } else if(started) {
+                    // A span was started but the current pixel is not to be included in the new SpanSet,
+                    // the current SpanSet should be close and added to the vector
+                    newVec.push_back(Span(y, minX, maxX));
+                    started = false;
+                }
+            }
+        }
+    }
+    return std::make_shared<SpanSet>(std::move(newVec));
+}
```

Comment:
I would need to think about it, but this may not be dealing with disconnected spans well. I think this is basically the first implementation I had but needed to expand things a bit. I do not favor getting rid of the boolean just to be cleaver as I think the bool helps with readability of the algorithm